### PR TITLE
fix duplicate isue

### PR DESCRIPTION
**Pull request by PavelD**
## Proposed change

I tried to address issue when the container is recreated.

## Type of change

- [x] Bugfix
- [ ] New feature
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Documentation

## Additional information

The new entity has been created every time docker recreate image. I decided to change definition of the unique id. 
The unique id is now depended on the `Compose stack`, `Compose service`, `Environment` and `config_entry.entry_id`.

The related issue: https://github.com/tomaae/homeassistant-portainer/issues/7

## Checklist

- [x] The code change is tested and works locally.
- [x] The code has been formatted using Black.
- [ ] Tests have been added to verify that the new code works.
- [ ] Documentation added/updated if required.

### CODE REVIEW

**Comment by PavelD** [comment]
Bump

**Comment by tomaae** [comment]
I'm bit confused here. Is "data_reference", container "Id" not unique?

**Comment by nvchambers** [comment]
> I'm bit confused here. Is "data_reference", container "Id" not unique?

Following the PR as I have the same issue.

- Container "ID"s within docker are unique **but not 'permanent'**, a new UUID is assigned after actions like a new image pull or container recreation.  Container names are persistent but IDs are ephemeral

- Screenshot of muliple HA entities generated by the Portainer service for one of my containers.  Entity detail in attached.

image

- Currently unable to use the HA portainer entities for automation as the random IDs generated by docker lead to new entities in HA.

Docker documentation suggests using a unique name for referencing because of the changing IDs:

[https://docs.docker.com/engine/containers/run/#container-identification](url)

_Defining a name can be a handy way to add meaning to a container. If you specify a name, you can use it when referring to the container in a user-defined network_

Appreciate your integration and have not yet tested the PR but wanted to confirm the behaviour @PavelD is trying to solve for

[portainer entity examples.txt](https://github.com/user-attachments/files/18453508/portainer.entity.examples.txt)

**Comment by tomaae** [comment]
I see, can you share debugs from integration? containers query is enough. So we can see what is available to us.

**Comment by nvchambers** [comment]
> I see, can you share debugs from integration? containers query is enough. So we can see what is available to us.

[portainer_debug_containers.json](https://github.com/user-attachments/files/18467183/portainer_debug_containers.json)

I can share more containers if needed - 3 from the portainer debug json attched

3 fields below might not be present for every integration user, relevant to using docker-compose vs docker run to create containers

 "Compose_Stack": 
 "Compose_Service": 
"Compose_Version": 

Thanks

**Comment by PavelD** [comment]
Sorry, i was not online for a while...

The integration is tracking state of services created by docker compose. The ID is generated randomly by docker engine and it's not persistent over restarts if the `--force-recreate` is used so i tried to use kind a constants defined in the portainer / docker compose file.

I was testing it on the most simple docker standalone environment only. I don't have more instances or more fancy setup like swarm, kubernetes or azure for tests. But generally the idea is to extend unique_id with more data if they are not unique for some configuration.

**Comment by PavelD** [comment]
PR is currently using `"Compose stack", "Compose service",` and `"Environment"`. It is fine for the standalone docker configuration. It is unique within the 1 instance of Portainer even if there is several environments pointing to the same docker.  It can clash if there is 2nd instance with the same setup.

I don't know how it is in other kind of environments.

**Comment by tomaae** [comment]
Yea, thats what I mean. If you have for example multiple nginx deploys in one environment. That is not unusual setup.
Endpoint id has to be there for sure tho.
Question is, does given container name have to be unique to one endpoint?

**Comment by nvchambers** [comment]
> Yea, thats what I mean. If you have for example multiple nginx deploys in one environment. That is not unusual setup. Endpoint id has to be there for sure tho. Question is, does given container name have to be unique to one endpoint?

TLDR - I believe "Name", "Environment" are the best options 
- Container Name is unque per environment, "Environment" name is easily updated in Portainer, "EndpointID" might complicate for those adding multiple Portainer hosts to to HA

**I am also not using Portainer to the full extent.  Can confirm a couple of things from my basic usage:**

- Container Name MUST be unique within an endpoint (it is effectively a network hostname so unique per endpointt)
    If attempting to create a container with a duplicate name it will be rejected or in some scenarios docker will add the random id to the front of the newly created overlapping name ie 34598c41ab6b_test for a container named test

- "Compose_Stack", "Compose_Service", "Compose_Version" are blank for Standalone containers within an endpoint
- "EndpointID" appears to be incremented for each new Environment you connect to Portainer, edittable by API
- "Environment" is easily configurable via the admin frontend

So:

One portainer host in the HA integration with multiple endpoints: EndpointID and name would be a unique combination
HostA -
Endpoint1 nginx = 1.nginx
Endpoint2 nginx = 2.niginx

Two portainer hosts in the HA integration: EndpointId and name might not be unique
Host A -
Endpoint1 nginx
Host B -
Endpoint1 nginx

Here A_1.nginx and B_1.nginx are not the same but the integration would not know

I am well outside of my usage but portainer docs support my thinking: https://docs.portainer.io/admin/environments/add/api

There is a note about the Id being used to make docker queries.
"Make a note of the Id value. It will be used to execute queries against the Docker Engine for the environment."

**Comment by tomaae** [comment]
Thanks for that.
Two portainer hosts would have different HA instance id, so that would not be a problem. Duplicate is only problem within one Portainer HA instance.

So endpointid and name will result in unique deviceid

**Comment by PavelD** [comment]
New version of PR:
The unique_id is now generated based on the docker image and environment - no dependency on docker compose.

I fix as well a few other issues:
  * icons are propagated to the sensor correctly
  * several environments defined on the portainer side are correctly reflected in HA
  * sensor naming is following environment names in portainer (example: Portainer local2 pihole )
  * Docker device is created for each portainer environment (Portainer local, Portainer local2)

But I see there still some issues. I need to double check if all changes are correct.

**Comment by PavelD** [comment]
Hi,
finally I had time to complete this PR. Now all data are stored per configuration.

The unique_id is composed from environment, name and configuration_id.

+ changes mentioned in my last comment.

What is not done: removal of stale entities but don't think it's needed for now.

BTW during the refactoring I removed following condition as it simplify the code and I don't know much why it was written in such complicated way. So I'm pointing this out.
```
if entity is None or (
  (entity_id not in platform.entities) and (entity.disabled is False)
):```

**Comment by jonwaland** [comment]
@PavelD   "What is not done: removal of stale entities but don't think it's needed for now."

I agree its much lower priority than fixing the error , but be aware that on a system with lots of containers, it builds up quickly...
image


If useful, cut down debug from my system attached.
[config_entry-portainer-0dc2aa38a4801df9b7ceb2adb6e7761f.json](https://github.com/user-attachments/files/18779044/config_entry-portainer-0dc2aa38a4801df9b7ceb2adb6e7761f.json)

**Comment by PavelD** [comment]
@jonwaland The PR is too large already. I think removal of stale containers should be in (not so big) separated PR.

**Comment by PavelD** [comment]
@jonwaland just for the curiosity. If you switch to debug and check logs the last log before it turns to be stale is `Error while updating entity + unique_id`?
I hope so.

**Comment by jonwaland** [comment]
> @jonwaland just for the curiosity. If you switch to debug and check logs the last log before it turns to be stale is `Error while updating entity + unique_id`? I hope so.

you want me to enable debug, update a container and check the log? sure. will try tonight.

**Comment by jonwaland** [comment]
using the current mainline version of the integration - unpatched. [debug.txt](https://github.com/user-attachments/files/18780459/debug.txt)

**Comment by PavelD** [comment]
@jonwaland Hi, sorry. I wrongly expected you are running the dev branch (with my changes). The way how to create/update sensors has been changed significantly. Let's wait id ti will be merged to the master successfully.

**Comment by b1ggi** [comment]
Im waiting eagerly, hope this finally ends:
image

**Comment by PHeonix25** [comment]
🤞 for a merge 🙏

**Comment by PavelD** [comment]
Soon we will soon celebrate 🎂 3 months since the creation of the PR. Any advice how to celebrate 🎉?

**Comment by jonwaland** [comment]
bump

**Comment by jonwaland** [comment]
bump

**Comment by supleed2** [comment]
These changes seemed to work for me too! (Thanks @PavelD ❤️)

A HACS release would be best probably, but for myself I went through and made a release (and bumped to v2 so I can keep track of it) which is usable by adding a "Custom repository" in HACS, [v2.0.1 here](https://github.com/supleed2/homeassistant-portainer/releases/tag/v2.0.1)

I don't know if it needs any maintenance of anything, but I'd be happy to try out / merge fixes or PRs on this repo too

**Comment by PavelD** [comment]
bump

**Comment by the-gangster** [comment]
> These changes seemed to work for me too! (Thanks @PavelD ❤️)
> 
> A HACS release would be best probably, but for myself I went through and made a release (and bumped to v2 so I can keep track of it) which is usable by adding a "Custom repository" in HACS, [v2.0.1 here](https://github.com/supleed2/homeassistant-portainer/releases/tag/v2.0.1)
> 
> I don't know if it needs any maintenance of anything, but I'd be happy to try out / merge fixes or PRs on this repo too

Thank you so much (to all that contributed to it as well). 
Almost missed it while waiting for a newer version to appear within HACS which would finally really bring the health states (mentioned in the readme) and the duplicate fix.
Just "updated" to your custom 2.0.1 version now and am lucky for now.

Only what I would like to mentioned:
When switching between the two, all devices and entities got duplicated and I found no way to remove the "old" devices any more. So I ended up removing both completely again and reintegrate the new one once more.
So who wants to preserve history or devices for automation purposes might better wait until the master repository gets updated.

**Comment by b1ggi** [comment]
Bump for master update
